### PR TITLE
Reclaim canisters

### PR DESCRIPTION
**Pull request by komal-rs**

### CODE REVIEW

File: src/canister/user_index/src/api/canister_management/recycle_canisters.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+        // update recycle_status
+
+        let success_canisters = results
+            .iter()
+            .filter(|r| r.is_ok() && r.as_ref().unwrap().is_some())
+            .map(|r| r.as_ref().unwrap().unwrap())
+            .collect::<Vec<Principal>>();
+
+        let num_success = success_canisters.len();
```
Comment (ravibazz) [inline src/canister/user_index/src/api/canister_management/recycle_canisters.rs:63]: Instead of this, we can use `run_task_concurrently` method from utils.

File: src/canister/user_index/src/api/canister_management/recycle_canisters.rs [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        None,
+        individual_user_template_canister_wasm.version.clone(),
+        individual_user_template_canister_wasm.wasm_blob.clone(),
+    )
+    .await
+    {
+        Ok(_) => Ok(Some(canister_id)),
+        Err(e) => Err((canister_id, e)),
+    }
+}
```
Comment (ravibazz) [inline src/canister/user_index/src/api/canister_management/recycle_canisters.rs:118]: This is same as `reset_user_individual_canisters`. See if we can refactor some code and use common functions.

File: src/canister/user_index/src/api/user_record/get_user_canister_list.rs [hunk 3]
```diff
@@ -16,6 +16,18 @@ fn get_user_canister_list() -> Vec<Principal> {
...
+fn get_user_id_and_canister_list() -> Vec<(Principal, Principal)> {
+    CANISTER_DATA.with(|canister_data_ref_cell| {
+        canister_data_ref_cell
+            .borrow()
+            .user_principal_id_to_canister_id_map
+            .iter()
+            .map(|(principal_id, canister_id)| (*principal_id, *canister_id))
+            .collect()
+    })
+}
```
Comment (ravibazz) [inline src/canister/user_index/src/api/user_record/get_user_canister_list.rs:29]: Why is this required?

File: src/canister/user_index/src/util/canister_management.rs [hunk 4]
```diff
@@ -94,52 +113,108 @@ pub async fn install_canister_wasm(canister_id: Principal, profile_owner: Option
...
+        canister_id,
+        wasm_module: wasm,
+        arg,
+    })
+    .await
+    {
+        Ok(_) => Ok(canister_id),
+        Err(err) => Err(err.1),
+    }
+}
```
Comment (ravibazz) [inline src/canister/user_index/src/util/canister_management.rs:157]: See if we can reuse this in `reset_user_individual_canisters`

File: src/canister/individual_user_template/src/data_model/mod.rs [hunk 5]
```diff
@@ -58,7 +67,9 @@ pub struct CanisterData {
     #[serde(default)]
     pub session_type: Option<SessionType>,
     #[serde(default)]
-    pub last_access_time: Option<SystemTime>
+    pub last_access_time: Option<SystemTime>,
+    #[serde(default)]
+    pub last_canister_functionality_access_time: Option<SystemTime>,
```
Comment (ravibazz) [inline src/canister/individual_user_template/src/data_model/mod.rs:72]: Why can't we use `last_access_time`. It was added for same purpose.

Comment (komal-rs) [reply src/canister/individual_user_template/src/data_model/mod.rs:72]: wanted to differentiate between last_user_access_time and last_canister_access_time

File: src/canister/user_index/src/api/user_record/get_user_canister_list.rs (hunk 3 above)
Comment (komal-rs) [reply src/canister/user_index/src/api/user_record/get_user_canister_list.rs:29]: Offchain agent gets a list of (user_id,canister_id). Filters acc to last_canister_access_time, fires call to user index to recycle and calls yral-metadata with user_id to remove keys

File: src/canister/user_index/src/api/canister_management/recycle_canisters.rs (hunk 1 above)
Comment (komal-rs) [reply src/canister/user_index/src/api/canister_management/recycle_canisters.rs:63]: yeah saw that, but I want to return the result and do some actions. using result_callback seemed less intuitive for that

File: src/canister/platform_orchestrator/Cargo.toml [hunk 6]
```diff
@@ -12,11 +12,12 @@ crate-type = ["cdylib"]
 [dependencies]
 candid = { workspace = true }
 ic-cdk = { workspace = true }
+ic-cdk-timers = { workspace = true }
```
[the remark this answers is not in the document]
Comment (komal-rs) [reply src/canister/platform_orchestrator/Cargo.toml:15]: will remove

File: src/canister/user_index/src/api/canister_management/mod.rs [hunk 7]
```diff
@@ -1,107 +1,233 @@
...
+        canister_id,
+        None,
+        individual_user_template_canister_wasm.version.clone(),
+        individual_user_template_canister_wasm.wasm_blob.clone(),
+    )
+    .await
+    {
+        Ok(_) => Ok(canister_id),
+        Err(e) => Err((canister_id, e)),
+    }
```
Comment (ravibazz) [inline src/canister/user_index/src/api/canister_management/mod.rs:232]: After recycling I think we should remove all the additional cycles that are on top of the base 0.2T cycles?

File: src/canister/individual_user_template/src/api/canister_lifecycle/post_upgrade.rs [hunk 8]
```diff
@@ -37,6 +21,8 @@ fn post_upgrade() {
     restore_data_from_stable_memory();
     save_upgrade_args_to_memory();
     reenqueue_timers_for_pending_bet_outcomes();
+    // TODO: to be removed on the next updagrade
+    update_last_canister_functionality_access_time();
```
Comment (ravibazz) [inline src/canister/individual_user_template/src/api/canister_lifecycle/post_upgrade.rs:25]: lets not depend on `post_upgrade`. Lets add trigger functions in platform_orchestrator and user_index so that we are not stuck to remove this before running any other emergency upgrades.

File: src/canister/user_index/src/api/user_record/get_user_canister_list.rs (hunk 3 above)
Comment (ravibazz) [reply src/canister/user_index/src/api/user_record/get_user_canister_list.rs:29]: okay so do we still require `get_user_canister_list`

File: src/canister/user_index/src/api/canister_management/mod.rs (hunk 7 above)
Comment (komal-rs) [reply src/canister/user_index/src/api/canister_management/mod.rs:232]: Any existing method I can use ?

File: src/canister/user_index/src/api/user_record/get_user_canister_list.rs (hunk 3 above)
Comment (komal-rs) [reply src/canister/user_index/src/api/user_record/get_user_canister_list.rs:29]: Yeah, being used for getting list of canisters for metrics collection

**Review by ravibazz — APPROVED** [review]

**Review by ravibazz — APPROVED** [review]
